src/buildstream/source.py

```diff
@@ -262,10 +262,123 @@ def __init__(
 
 @dataclass
 class AliasSubstitution:
+    """AliasSubstitution()
+    An opaque data structure which may be passed through
+    :func:`SourceFetcher.fetch() <buildstream.source.SourceFetcher.fetch>` and in such cases
+    must be provided to :func:`Source.translate_url() <buildstream.source.Source.translate_url>`.
+    """
+
     _effective_alias: str
     _mirror: Union[SourceMirror, str]
 
 
+class SourceInfoMedium(FastEnum):
+    """
+    Indicates the meduim in which the source is obtained
+
+    *Since: 2.5*
+    """
+
+    LOCAL = "local"
+    """
+    Files stored locally in the project
+    """
+
+    ARCHIVE = "archive"
```

Comment:
`LOCAL` vs. `ARCHIVE` seem like orthogonal attributes to me. You can have a local archive in your project and you can fetch a single e.g. config file from a remote server, at least theoretically. Maybe `LOCAL` should rather be a flag than part of this enum? (Or maybe even make it implicit as the corresponding URL will have no host component).

I'm also not sure whether we should have one enum entry for each archive file format (e.g. `TAR` and `ZIP`) or whether a generic `ARCHIVE` makes more sense.

Reply:
Interesting, I consider the *medium* type as the *means in which the data is stored*, and not what the downloaded data *actually is*.

I.e. the medium here explains that it comes from the project itself, while a git medium used a `git` protocol to download, and `archive` downloaded a file.

I rather lean the other way here and think that `ARCHIVE` may be renamed to `REMOTE_FILE` or such, and this would make sense for any `DownloadableFileSource`.

When considering this, I'm trying  to balance:
* What is useful for interoperability with external tooling
* What is needed to logically describe the provenance of the source, for reporting purposes

Reply:
I've now renamed this to be `REMOTE_FILE` which is used by default by `DownloadableFileSource`, and is, in my opinion factually correct for any remotely downloaded file, including possibly compressed archives.